docker/Dockerfile

```diff
@@ -0,0 +1,111 @@
+RUN tar -xzf gcloud-sdk.tar.gz
+RUN ./google-cloud-sdk/install.sh --quiet
+RUN rm gcloud-sdk.tar.gz
+
+WORKDIR /app/oppia
+
+# installing python dependencies from the requirements_dev.txt file
+COPY requirements_dev.in .
+COPY requirements_dev.txt .
+
+RUN pip-compile --generate-hashes requirements_dev.in
+RUN pip install --require-hashes --no-deps -r requirements_dev.txt
+
+## installing packages from the package.json file
+COPY package.json .
+COPY scripts/linters/custom_eslint_checks ./scripts/linters/custom_eslint_checks
+
+RUN npm install -g yarn
+# mathjs ESOCKETTIMEDOUT error fix -- https://stackoverflow.com/questions/55845756/esockettimedout-error-while-yarn-install-command
+RUN yarn install --network-timeout 1000000
+
+# installing buf and proto for Linux -- this docker container is based on Linux
+COPY buf.gen.yaml .
+COPY /extensions/classifiers/proto/ ./extensions/classifiers/proto/
+
+# installing third party dependencies
+COPY scripts ./scripts
+COPY /core ./core
+COPY /assets ./assets
+COPY dependencies.json .
+
+# installation for proto buf requires the third party dependencies to be installed (oppia-ml-proto)
+RUN python -m scripts.install_third_party
+
+ENV BUF_LINUX_FILES="buf-Linux-x86_64 protoc-gen-buf-check-lint-Linux-x86_64 protoc-gen-buf-check-breaking-Linux-x86_64" \
+    PROTOC_LINUX_FILE='protoc-3.13.0-linux-x86_64.zip' \
+    BUF_DIR='/app/buf-0.29.0' \
+    PROTOC_DIR='/app/buf-0.29.0/protoc' \
+    BUF_BASE_URL='https://github.com/bufbuild/buf/releases/download/v0.29.0/' \
+    PROTOC_URL='https://github.com/protocolbuffers/protobuf/releases/download/v3.13.0'
+
+# Download BUF_LINUX_FILES and PROTOC_LINUX_FILE
+# set the shell as /bin/bash explicitly. This is required to make use of the array syntax correctly
+SHELL ["/bin/bash", "-c"]
+RUN IFS=' ' read -ra BUF_FILES <<< "$BUF_LINUX_FILES"; \
+    for bin_file in "${BUF_FILES[@]}"; do \
+        wget -P $BUF_DIR $BUF_BASE_URL/$bin_file; \
+    done
+
+RUN wget -P $BUF_DIR $PROTOC_URL/$PROTOC_LINUX_FILE
+RUN unzip $BUF_DIR/$PROTOC_LINUX_FILE -d $PROTOC_DIR \
+    && rm $BUF_DIR/$PROTOC_LINUX_FILE
+
+RUN chmod -R 744 $BUF_DIR \
+    && chmod -R 744 $PROTOC_DIR
+
+# compiling the protobuf files
+ENV PROTOC_FILES_PATH='/app/oppia/third_party/oppia-ml-proto-0.0.0'
+
+RUN mkdir /app/oppia/proto_files
+RUN /app/buf-0.29.0/protoc/bin/protoc \
+    --plugin=protoc-gen-ts=/app/oppia/node_modules/.bin/protoc-gen-ts \
+    --ts_out=/app/oppia/extensions/classifiers/proto/ \
+    --js_out=import_style=commonjs,binary:/app/oppia/extensions/classifiers/proto/ \
+    --python_out=/app/oppia/proto_files \
+    --proto_path=$PROTOC_FILES_PATH \
+    $PROTOC_FILES_PATH/*.proto
+
+RUN sed -i 's/import text_classifier_pb2 as text__classifier__pb2/from . import text_classifier_pb2 as text__classifier__pb2/' /app/oppia/proto_files/training_job_response_payload_pb2.py
+
+RUN python -m scripts.build
```

Comment:
for generating the `third_party/generated` in the dev_mode

Reply:
as per discussion with @U8NWXD, the build.py specifically in dev_mode is used to generate `third_party/generated` dir which is a unified code for the third_party packages that are being downloaded using `dependencies.json`. as we are moving forward with the depedencies.json deprecation, there will not be any addition of the packages in this json file. so we can add this step in the dockerfile so it gets in the cache.

Reply:
Thanks for looking into it in detail, @shivkant5 

If it's just dealing with dependencies.json libs, can we move those specific steps to `install_dependencies_json_packages.py`, so that we only have to use this single/temporary/independent file for the image and it can be easily removed when needed?

Reply:
but this is not just we are doing this thing only when in dev_mode. we clean the existing generated files, build some hashes files.
when in prod_mode, minify the codebase, generate backend prod_files and other prod related files, generate app.yml file specific for prod
within minify flag (this flag is used internally within the scripts), we minify the third_party_libs.

so it seems we cannot move this toi install_dependencies_json file.
what I want to ask from you is, we clean the setup and re-generate the `thirdparty/generated` dir, means we are refreshing it everytime we start the server. what do you say, should we add this in the compose service command? so that it get refreshed everytime we start the server. I assume it wont take more than 30-40secs to regenerate

Reply:
> can you please list out the thing for which we are using build.py?

This is why I'm asking to create a list. It will be hard to decide otherwise.

> . we clean the existing generated files

This is not needed if we are doing it in the image layer.

> build some hashes files.

Do we need to create this before starting the server with new files? i.e, part of entry point?

> when in prod_mode, minify the codebase, generate backend prod_files and other prod related files, generate app.yml file specific for prod
within minify flag (this flag is used internally within the scripts), we minify the third_party_libs.

So it looks like we can't use most of the steps in build.py  during image creation phase as it depends on local repo.


> so it seems we cannot move this toi install_dependencies_json file.

Reading the above comment, I feel like we should move the dependencies.json related things to the install_dependencies_json.py file as will help us move the static part of build.py to Dockerfile.

Reply:
This is why I'm asking to create a list. It will be hard to decide otherwise.
-- in dev_mode - 1) clean generated files 2) re-generate these third party files 3) generate some hashes file. this is for dev mode only when no flag is passed. (when passing prod_env I am thinking to add this in compose service command as it will generate the prod artifacts when prod env flag passed)

Do we need to create this before starting the server with new files? i.e, part of entry point?
-- yes the hashes files need to be created before the server is up

So it looks like we can't use most of the steps in build.py during image creation phase as it depends on local repo.
-- yes, we need to use them while starting the container

I would suggest leaving the third_party/generated part in the Dockerfile itself via install_dependencies_json.py file and move other dynamic steps to entrypoint.sh. wdyt?
-- I am okay with this, but one point I need to clarify, we will create the third_party/generated folder within the `install_dependencies_json_packages`, and the dynamic part will be moved to the compose entrypoint. but the problem is that the build.py will first execute the clean which will delete all the generated file (which we generate within the image building step and we cannot remove this step). so this will be clearing the generated dirs

Reply:
Let's do this:
```
- Dockerfile:
    - Generates `third_party/generated` folder
    - Creates empty `{}` hashes.json
- entrypoint.sh
    - Modify contant
    - if prod_mode: run build.py to do all the prod related things.
```

This structure will help us (1) simplify the dev_mode and make it faster to start devserver (2) Helps us avoid the OPPIA_IS_DOCKERISED flag in build.py

Let me know if this sounds good to you?